src/coreclr/System.Private.CoreLib/src/System/Runtime/CompilerServices/AsyncHelpers.CoreCLR.cs

```diff
@@ -171,240 +203,386 @@ private static unsafe object AllocContinuationResultBox(void* ptr)
+
+                            return;
+                        }
+
+                        nextContinuation.SetException(ex);
+
+                        continuation = nextContinuation;
+                    }
+
+                    if (continuation.Resume == null)
+                    {
+                        bool successfullySet = TOps.SetCompleted(task, continuation);
+
+                        contexts.Pop();
+
+                        if (!successfullySet)
+                        {
+                            ThrowHelper.ThrowInvalidOperationException(ExceptionResource.TaskT_TransitionToFinal_AlreadyCompleted);
+                        }
+
+                        return;
+                    }
+
+                    if (QueueContinuationFollowUpActionIfNecessary<T, TOps>(task, continuation))
+                    {
+                        contexts.Pop();
+                        return;
                     }
                 }
             }
-        }
 
-        private static async ValueTask FinalizeValueTaskReturningThunk(Continuation continuation)
-        {
-            Continuation finalContinuation = new Continuation
+            private static Continuation UnwindToPossibleHandler(Continuation continuation)
             {
-                Flags = CorInfoContinuationFlags.CORINFO_CONTINUATION_NEEDS_EXCEPTION,
-            };
-            continuation.Next = finalContinuation;
+                while (true)
+                {
+                    Debug.Assert(continuation.Next != null);
+                    continuation = continuation.Next;
+                    if ((continuation.Flags & CorInfoContinuationFlags.CORINFO_CONTINUATION_NEEDS_EXCEPTION) != 0)
+                        return continuation;
+                }
+            }
 
-            while (true)
+            public static void HandleSuspended<T, TOps>(T task) where T : Task where TOps : IThunkTaskOps<T>
             {
-                Continuation headContinuation = UnlinkHeadContinuation(out var awaitableProxy);
-                await awaitableProxy;
-                Continuation? finalResult = DispatchContinuations(headContinuation);
-                if (finalResult != null)
+                Continuation headContinuation = UnlinkHeadContinuation(out INotifyCompletion? notifier);
+
+                // Head continuation should be the result of async call to AwaitAwaiter or UnsafeAwaitAwaiter.
+                // These never have special continuation handling.
+                const CorInfoContinuationFlags continueFlags =
+                    CorInfoContinuationFlags.CORINFO_CONTINUATION_CONTINUE_ON_CAPTURED_SYNCHRONIZATION_CONTEXT |
+                    CorInfoContinuationFlags.CORINFO_CONTINUATION_CONTINUE_ON_THREAD_POOL |
+                    CorInfoContinuationFlags.CORINFO_CONTINUATION_CONTINUE_ON_CAPTURED_TASK_SCHEDULER;
+                Debug.Assert((headContinuation.Flags & continueFlags) == 0);
+
+                // TODO: This may need volatile write
```

Comment:
What writes we would be ordering? Is there a volatile read that this would pair with?

Reply:
This sets `Task.m_stateObject = continuation` and then calls `UnsafeOnCompleted` on an arbitrary awaiter. The awaiter may then resume the continuation on a different thread, which will end up in `ThunkTaskCore.MoveNext` that reads `Task.m_stateObject` in the beginning. The volatile read would be that read of `Task.m_stateObject`.

I guess there is some kind of data dependency, but I am not sure if that is sufficient or not. There is similar concern with when we move to a different `SynchronizationContext`/`TaskScheduler` below. What do you think?

Reply:
That is equivalent to assigning an object after initializing all its fields to some statically accessible variable and then using that object and its fields from a different thread. That does not require additional ordering. 
Object assignments have release semantic related to the object's fields. Combined with data dependency on the reading side it is enough for the "exchange via shared state" to be consistent.

Reply:
Great, I will just remove the TODOs, then. Thanks!